### PR TITLE
wails: fix build on darwin

### PR DESCRIPTION
**Pull request by afh**
## Description of changes

* Replace `stdenv` with `gccStdenv`
* Replace `gcc` with `stdenv.cc.cc`
* Add `buildInputs` depending on platform
* ~~Add `meta.mainProgram`~~ already added by ff1a94e523ae9fb272e0581f068baee5d1068476

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by ianmjones** [comment]
Result of `nixpkgs-review pr 293576` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>snippetexpandergui</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>wails</li>
  </ul>
</details>

**Comment by afh** [comment]
Thanks for testing on linux, @ianmjones, much appreciated! Unfortunately I do not have access to a linux machine at the moment and the passthru tests on here on GitHub were successful.

Could you share some details or logs from your `nixpkgs-review` run?

**Comment by ianmjones** [comment]
Hey @afh,

Thanks for adding support for darwin, I don't have the ability to test on nix-darwin, so appreciate it.

In fact, would you mind adding yourself as a maintainer of the package so that someone who uses Wails on macOS can give any future releases a test?

nixpkgs-review gave the following output:

2024-03-05T23:35:28,157362452+00:00

Afterwards, I installed `wails` from that build, and tested building `snippetexpandergui` locally (I'm the author of that software), and creating and building a brand new app with `wails`, worked fine. :shrug:

**Comment by afh** [comment]
@ianmjones could you run `nixpkgs-review pr 293576` on x86_64-linux once again to confirm that the latest changes fix the `snippetexpandergui` [build failure reported earlier](https://github.com/NixOS/nixpkgs/pull/293576#issuecomment-1979805711)?

**Comment by ianmjones** [comment]
Result of `nixpkgs-review pr 293576` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>snippetexpandergui</li>
    <li>wails</li>
  </ul>
</details>

**Review by ianmjones — APPROVED** [review]
Yay, works well! :+1:

File: pkgs/development/tools/wails/default.nix [hunk 1]
```diff
@@ -58,18 +59,18 @@ buildGoModule rec {
   # As Wails calls a compiler, certain apps and libraries need to be made available.
   postFixup = ''
     wrapProgram $out/bin/wails \
-      --prefix PATH : ${lib.makeBinPath [ pkg-config go gcc nodejs ]} \
-      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ gtk3 webkitgtk ]} \
+      --prefix PATH : ${lib.makeBinPath [ pkg-config go stdenv.cc nodejs ]} \
+      --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath ([ ] ++ lib.optionals stdenv.isLinux [ gtk3 webkitgtk ])}" \
```
Comment (kirillrdy) [inline pkgs/development/tools/wails/default.nix:63]: don't need  `[ ]++`  lib.optionals returns `[ ]` when first argument is false

Comment (afh) [reply pkgs/development/tools/wails/default.nix]: Thanks, @kirillrdy, that's good to know 👍 I've update this PR accordingly.

**Comment by ianmjones** [comment]
Result of `nixpkgs-review pr 293576` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>snippetexpandergui</li>
    <li>wails</li>
  </ul>
</details>

**Review by ianmjones — APPROVED** [review]

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -20053,7 +20053,9 @@ with pkgs;
 
   vtable-dumper = callPackage ../development/tools/misc/vtable-dumper { };
 
-  wails = callPackage ../development/tools/wails { };
+  wails = callPackage ../development/tools/wails {
+    stdenv = gccStdenv;
+  };
```
Comment (wegank) [inline pkgs/top-level/all-packages.nix:20058]: Is this really necessary on darwin?

Comment (afh) [reply pkgs/top-level/all-packages.nix:20058]: Thanks for asking, @wegank. When taking this on I did not want to switch the darwin build from gcc to clang without further investigation.
Yet local tests show that with `stdenv = if stdenv.isLinux then gccStdenv else stdenv;` wails also builds on aarch64-darwin. Although I haven't done any extensive testing with the resulting binary.
Anything in particular that would be worthwhile to test/scrutinize?

/ccc @ianmjones

Comment (ianmjones) [reply pkgs/top-level/all-packages.nix:20058]: FYI, I tested without `stdenv = gccStdenv;` on x86_64-darwin, and apps would not compile with `wails`, not even the tiny default app created with the `svelte` template.

Comment (wegank) [reply pkgs/top-level/all-packages.nix:20058]: Ah, thanks for the info. Let's leave it as is.

**Review by wegank — APPROVED** [review]

**Pull request merged**